### PR TITLE
Makes TextBoundary and its subclasses public

### PR DESCRIPTION
**Pull request by chunhtai**
Make it public so it can be reused in SelectableRegion

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/text_boundary.dart:14]: nit: "code unit" -> "UTF-16 code unit"

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/text_boundary.dart:14]: Actually I think as a public API it will be easier to interpret all `TextPosition`s as caret locations (since it has an affinity).

File: packages/flutter/lib/src/rendering/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/text_boundary.dart:8]: what's the reason this is put in rendering (as opposed to, for example, services)?

File: packages/flutter/lib/src/rendering/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/text_boundary.dart:46]: utf8 -> UTF-16

File: packages/flutter/lib/src/rendering/text_boundary.dart
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/text_boundary.dart:8]: services probably makes more sense, will do

File: packages/flutter/lib/src/painting/text_painter.dart
Comment (chunhtai) [inline packages/flutter/lib/src/painting/text_painter.dart:1041]: This should be in the ui.text, putting it here just for now, will spawn another pr and remove this before merge

File: packages/flutter/test/material/text_field_test.dart
Comment (chunhtai) [inline packages/flutter/test/material/text_field_test.dart:3375]: previous test is testing the implementation detail of textOffsetToPosition

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:61]: Is the `offset -1` strategy correct? If there's no bidi text or soft line wraps in a paragraph, then we have
```
TextPosition(x, downstream) == TextPosition(x, upstream) != TextPosition(x-1, downstream)
```
 right?

File: packages/flutter/lib/src/rendering/editable.dart
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/editable.dart:2085]: directionality should be fine as the skia should handle that for us? The problem I facing here is I can't differentiate whether user taps the upstream of the end of the text or far away to the right from the end of the text. I think the correct behavior is we should still select word for the former and collapse at the end for the latter.

Since we can't tell it with the current API, I made it always collapse at the end, since that is what the original behavior is

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (chunhtai) [reply packages/flutter/lib/src/services/text_boundary.dart]: I am not sure what description i should add here. That will be up to the application to interpret the output?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (chunhtai) [reply packages/flutter/lib/src/services/text_boundary.dart]: There is really nothing else i can explain in this constructor.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (chunhtai) [reply packages/flutter/lib/src/services/text_boundary.dart]: no you are right, I will make this private

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (chunhtai) [reply packages/flutter/lib/src/services/text_boundary.dart]: My understanding of codeUnitboundary is the following. If you have a text 'abc'. To describe the boundary of code unit 'b' will be

leading : TextPosition(offset: 1, affinity:TextAffinity.downStream)
trailing:  TextPosition(offset: 2, affinity:TextAffinity.upStream)

The current code should capture this. It make sure any text position that point to 'b' will return that. WDYT?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart:64]: If the text is `ABC` and assume the paragraph is LTR and there's no soft wrapping, then `TextPosition(1, downstream)` and `TextPosition(1, upstream)` point to exactly the same caret position visually (at least based on the logic in `TextPainter`), but this class will return a different character range for these 2 ('B' with the downstream one and 'A' with the upstream one). I think it could be confusing to the app users that the same caret location (at least that's what appears to be the case for them) actually points to different characters.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart]: Yeah according to what's in `TextPainter` that looks correct

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart:64]: The "right" behavior to me here, if we want to account for text affinity, is probably to return the same range when `TextPosition(1, downstream)` and `TextPosition(1, upstream)` both give the same caret position. We'll have to query the text layout in order to do that.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart:64]: Discussed offline, `TextPosition(offset: x, affinity: upstream)` should be interpreted as pointing to the trailing half of the `x-1`th code unit, and `TextPosition(offset: x, affinity: downstream)` is the leading half of the `x`th code unit. Since this applies to all `TextBoundary` subclasses maybe we should explain that in `TextBoundary`'s documentation?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:129]: shouldn't this be `upstream`, following `LineBreak`'s logic? Or should the word breaker behave differently?

**Review by LongCatIsLooong — COMMENTED** [review]
LGTM with nits/questions

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:4327]: do we still need this, now that we always treat `TextPosition` as caret locations?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart:4327]: Yes otherwise you won't be able to move to the next word if the current position is at the edge. This boundary move the position by one affinity. This may need a better name though...

**Comment by chunhtai** [comment]
blocked on https://github.com/flutter/engine/pull/36176/files

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (chunhtai) [reply packages/flutter/lib/src/services/text_boundary.dart:64]: done

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (chunhtai) [reply packages/flutter/lib/src/services/text_boundary.dart:129]: nice catch, will fix

**Review by LongCatIsLooong — APPROVED** [review]

**Pull request merged**
